engine/sync_manager.go

```diff
@@ -397,72 +356,72 @@ func (m *syncManager) Update(exchangeName string, p currency.Pair, a asset.Item,
 	}
 
-	m.mux.Lock()
-	defer m.mux.Unlock()
+	k := currencyPairKey{
+		AssetType: a,
+		Exchange:  exchangeName,
+		Pair:      p,
+	}
 
-	for x := range m.currencyPairs {
-		if m.currencyPairs[x].Exchange == exchangeName &&
-			m.currencyPairs[x].Pair.Equal(p) &&
-			m.currencyPairs[x].AssetType == a {
-			switch syncType {
-			case SyncItemTicker:
-				origHadData := m.currencyPairs[x].Ticker.HaveData
-				m.currencyPairs[x].Ticker.LastUpdated = time.Now()
-				if err != nil {
-					m.currencyPairs[x].Ticker.NumErrors++
-				}
-				m.currencyPairs[x].Ticker.HaveData = true
-				m.currencyPairs[x].Ticker.IsProcessing = false
-				if atomic.LoadInt32(&m.initSyncCompleted) != 1 && !origHadData {
-					removedCounter++
-					log.Debugf(log.SyncMgr, "%s ticker sync complete %v [%d/%d].",
-						exchangeName,
-						m.FormatCurrency(p).String(),
-						removedCounter,
-						createdCounter)
-					m.initSyncWG.Done()
-				}
-				return nil
-			case SyncItemOrderbook:
-				origHadData := m.currencyPairs[x].Orderbook.HaveData
-				m.currencyPairs[x].Orderbook.LastUpdated = time.Now()
-				if err != nil {
-					m.currencyPairs[x].Orderbook.NumErrors++
-				}
-				m.currencyPairs[x].Orderbook.HaveData = true
-				m.currencyPairs[x].Orderbook.IsProcessing = false
-				if atomic.LoadInt32(&m.initSyncCompleted) != 1 && !origHadData {
-					removedCounter++
-					log.Debugf(log.SyncMgr, "%s orderbook sync complete %v [%d/%d].",
-						exchangeName,
-						m.FormatCurrency(p).String(),
-						removedCounter,
-						createdCounter)
-					m.initSyncWG.Done()
-				}
-				return nil
-			case SyncItemTrade:
-				origHadData := m.currencyPairs[x].Trade.HaveData
-				m.currencyPairs[x].Trade.LastUpdated = time.Now()
-				if err != nil {
-					m.currencyPairs[x].Trade.NumErrors++
-				}
-				m.currencyPairs[x].Trade.HaveData = true
-				m.currencyPairs[x].Trade.IsProcessing = false
-				if atomic.LoadInt32(&m.initSyncCompleted) != 1 && !origHadData {
-					removedCounter++
-					log.Debugf(log.SyncMgr, "%s trade sync complete %v [%d/%d].",
-						exchangeName,
-						m.FormatCurrency(p).String(),
-						removedCounter,
-						createdCounter)
-					m.initSyncWG.Done()
-				}
-				return nil
-			}
+	c, exists := m.currencyPairs[k]
+	if !exists {
+		return fmt.Errorf("%w for %s %s %s", errCouldNotSyncNewData, k.Exchange, k.Pair, k.AssetType)
+	}
+
+	c.locks[syncType].Lock()
+	defer c.locks[syncType].Unlock()
+
+	if c.trackers[syncType] == nil {
+		c.trackers[syncType] = &syncBase{}
+	}
+	s := c.trackers[syncType]
+
+	if !s.IsUsingWebsocket {
+		s.IsUsingWebsocket = true
+		s.IsUsingREST = false
+		n := "Ticker"
+		if syncType == SyncItemOrderbook {
+			n = "Orderbook"
+		} else if syncType == SyncItemTrade {
+			n = "Trade"
 		}
+		log.Warnf(log.SyncMgr,
+			"%s %s %s: %s Websocket re-enabled, switching from rest to websocket",
+			c.Exchange,
+			m.FormatCurrency(c.Pair),
+			strings.ToUpper(c.AssetType.String()),
+			n,
+		)
+	}
+
+	return m.update(c, syncType, err)
```

Comment:
With websocket enabled. `.\gocryptotrader -tickersync=true -orderbooksync=false`  Orderbooks are still being processed same for `.\gocryptotrader -tickersync=false -orderbooksync=true`. I think we can just return for now if sync type is disabled and add a TODO to limit subscriptions wrapper side.

Reply:
Need to grok why that's the case. 
342-354 should have been preventing that.

Reply:
Yep. I don't get why from the code.

 [342-354](https://github.com/thrasher-corp/gocryptotrader/blob/c906adfc22fa0eb77c531d5f03dd2de5464061d9/engine/sync_manager.go#L342-L354) should have guarded against that in calls to Update,
and direct internal calls to update should have been guarded by [504-512](https://github.com/thrasher-corp/gocryptotrader/blob/c906adfc22fa0eb77c531d5f03dd2de5464061d9/engine/sync_manager.go#L504-L512)

Will try to see why it's getting there.

Reply:
@shazbert Do you mean this:
`[DEBUG] | SYNC | 15/06/2023 18:23:29 | Okx orderbook sync complete BTC-USDT [1/2].`
Which I still get when running:
`go run main.go -tickersync=false -orderbooksync=true`

If so, I can fix it, but it was the same on master before, so I think can be done in a separate PR.

Reply:
If its on master yeah we can keep it the same. 👍 Push what you have worked on and I will send you a diff if I figure anything out. 🚀 Thanks heaps.